src/store/swapOrderSlice.ts

```diff
@@ -0,0 +1,197 @@
+              variant: 'info',
+            }),
+          )
+
+          break
+        case 'presignaturePending':
+          dispatch(
+            showNotification({
+              title: 'Order waiting for signature',
+              message: 'Waiting for confirmation from signers of your Safe',
+              groupKey,
+              variant: 'info',
+            }),
+          )
+          break
+        case 'open':
+          dispatch(
+            showNotification({
+              title: 'Order transaction confirmed',
+              message: 'Waiting for order execution by the CoW Protocol',
+              groupKey,
+              variant: 'info',
+            }),
+          )
+          break
+        case 'fulfilled':
+          if (oldStatus === undefined) {
+            return
+          }
+          dispatch(
+            showNotification({
+              title: 'Order executed',
+              message: 'Your order has been successful',
+              groupKey,
+              variant: 'info',
+            }),
+          )
+          break
+        case 'expired':
+          if (oldStatus === undefined) {
+            return
+          }
+          dispatch(
+            showNotification({
+              title: 'Order expired',
+              message: 'Your order has reached the expiry time and has become invalid',
+              groupKey,
+              variant: 'warning',
+            }),
+          )
+          break
+        case 'cancelled':
+          if (oldStatus === undefined) {
+            return
+          }
+          dispatch(
+            showNotification({
+              title: 'Order cancelled',
+              message: 'Your order has been cancelled',
+              groupKey,
+              variant: 'warning',
+            }),
+          )
+          break
+      }
+    },
+  })
+}
+
+/**
+ * Listen for changes in the tx history, check if the transaction is a swap order and update the status of the order
+ * @param listenerMiddleware
+ */
+export const swapOrderListener = (listenerMiddleware: typeof listenerMiddlewareInstance) => {
+  listenerMiddleware.startListening({
+    actionCreator: txHistorySlice.actions.set,
+    effect: (action, listenerApi) => {
+      if (!action.payload.data) {
+        return
+      }
+
+      for (const result of action.payload.data.results) {
+        if (!isTransactionListItem(result)) {
+          continue
+        }
+
+        if (isSwapTxInfo(result.transaction.txInfo)) {
+          const swapOrder = result.transaction.txInfo
+          const oldStatus = selectSwapOrderStatus(listenerApi.getOriginalState(), swapOrder.uid)
+
+          if (oldStatus === swapOrder.status) {
+            continue
+          }
+
+          listenerApi.dispatch({
+            type: slice.actions.setSwapOrder.type,
+            payload: {
+              orderUid: swapOrder.uid,
+              status: swapOrder.status,
+              txId: result.transaction.id,
+            },
+          })
+        }
```

Comment:
The `swapOrderListener` is very similar to the `txHistoryListener`. Anything against moving this part into the `txHistoryListener` instead of defining another one on the same slice?

Reply:
I wouldn't do this. I would prefer to keep them separated. I know that they are similar, but still distinct and I would like to avoid touching "non-swap" parts. This would also complicate the original middleware. 

I think that also this separation is ok, as you would have everything related to the swap at one place and won't be surprised to find logic related to it in an unrelated place.

Reply:
Agreed, after thinking about it it makes sense to have the slice as the source of truth and separate listeners by feature instead of having one listener that handles multiple things. Its much easier to look up that swap orders are affected by the history than having to know the context that the history is affecting swap orders.